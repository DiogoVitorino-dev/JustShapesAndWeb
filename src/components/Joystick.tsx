import React from "react";
import { View, StyleSheet } from "react-native";
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import Animated, {
  Easing,
  useAnimatedStyle,
  useSharedValue,
  interpolate,
  useAnimatedReaction,
  withTiming,
  runOnJS,
} from "react-native-reanimated";

interface JoyStickData {
  x: number;
  y: number;
  angle: number;
}

interface JoyStickProp {
  size?: number;
  velocity?: number;
  onMove: (data: JoyStickData) => void;
}

const calculateAngle = (x: number, y: number) => {
  "worklet";
  let angle = Math.atan2(y, x); // calculating angle
  angle = angle * (180 / Math.PI); // convert radian to degree
  angle = (angle + 360) % 360; // convert to positives degrees
  return angle;
};

export default function Joystick({
  size = 100,
  velocity = 100,
  onMove,
}: JoyStickProp) {
  const radius = size / 2;
  const pos = useSharedValue({ x: 0, y: 0 });

  const normalizeData = ({ angle, x, y }: JoyStickData): JoyStickData => {
    "worklet";
    const distance = Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2));

    x = interpolate(x, [distance * -1, distance], [velocity * -1, velocity]);
    y = interpolate(y, [distance * -1, distance], [velocity * -1, velocity]);

    return { angle, x, y };
  };

  const pan = Gesture.Pan()
    .onBegin((event) => {
      const posX = event.x - radius;
      const posY = event.y - radius;

      pos.value = { x: posX, y: posY };
    })

    .onUpdate((event) => {
      const posX = event.x - radius;
      const posY = event.y - radius;

      // Distance from center to the edge of the area
      const distance = Math.sqrt(Math.pow(posX, 2) + Math.pow(posY, 2));

      // Limits the pointer to the edge of the area
      if (distance >= radius) {
        const factor = radius / distance;
        pos.value = { x: posX * factor, y: posY * factor };
      } else {
        pos.value = { x: posX, y: posY };
      }
    })

    .onFinalize(() => {
      pos.value = { x: 0, y: 0 };
    })
    .withTestId("pan");

  const pointerAnimatedStyle = useAnimatedStyle(() => ({
    transform: [
      {
        translateX: withTiming(pos.value.x, {
          duration: 20,
          easing: Easing.out(Easing.ease),
        }),
      },
      {
        translateY: withTiming(pos.value.y, {
          duration: 20,
          easing: Easing.out(Easing.ease),
        }),
      },
    ],
  }));

  useAnimatedReaction(
    () => {
      const angle = calculateAngle(pos.value.x, pos.value.y);
      return normalizeData({ angle, x: pos.value.x, y: pos.value.y });
    },
    (currentValue, previousValue) => {
      if (currentValue !== previousValue) {
        runOnJS(onMove)(currentValue);
      }
    },
  );

  return (
    <View style={styles.container}>
      <GestureDetector gesture={pan}>
        <View
          testID="pan"
          style={[
            { width: size, height: size, borderRadius: size / 2 },
            styles.area,
          ]}
        >
          <Animated.View
            style={[
              { width: size / 2.5, height: size / 2.5, borderRadius: size / 2 },
              styles.pointer,
              pointerAnimatedStyle,
            ]}
          />
        </View>
      </GestureDetector>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
  },

  area: {
    backgroundColor: "tomato",
    alignItems: "center",
    justifyContent: "center",
  },

  pointer: {
    backgroundColor: "cyan",
  },
});
