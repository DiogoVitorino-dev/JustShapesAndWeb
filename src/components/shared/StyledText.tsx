import React from "react";
import { Text, StyleSheet, TextProps } from "react-native";

interface StyledTextProp extends TextProps {}

const StyledText = ({ children, style, ...others }: StyledTextProp) => (
  <Text {...others} style={[styles.default, style]}>
    {children}
  </Text>
);

export const MegrimText = ({ style, ...others }: StyledTextProp) =>
  StyledText({ style: [style, styles.fontMegrim], ...others });

export const MajorText = ({ style, ...others }: StyledTextProp) =>
  StyledText({ style: [style, styles.fontMajor], ...others });

export const ManjariText = ({ style, ...others }: StyledTextProp) =>
  StyledText({ style: [style, styles.fontManjari], ...others });

const styles = StyleSheet.create({
  default: {
    fontSize: 22,
    color: "white",
  },
  fontMegrim: { fontFamily: "Megrim" },
  fontManjari: { fontFamily: "Manjari" },
  fontMajor: { fontFamily: "MajorMonoDisplay" },
});
