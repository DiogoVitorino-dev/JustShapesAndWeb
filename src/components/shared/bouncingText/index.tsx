import React, { useMemo } from "react";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
} from "react-native-reanimated";

import Character, { CharacterProps } from "./character";

import { AnimatedProps, DisplayOptions } from "@/constants/commonTypes";

type ForwardCharacterProps = Omit<CharacterProps, "index" | "style">;

type BouncingTextStyle =
  | CharacterProps["style"]
  | ((characterIndex: number) => CharacterProps["style"]);

export interface BouncingTextProps extends ForwardCharacterProps {
  containerStyle?: AnimatedProps<"View">["style"];
  style?: BouncingTextStyle;
}

export function BouncingText({
  start,
  children,
  containerStyle,
  style,
  onFinish,
  ...props
}: BouncingTextProps) {
  const display = useSharedValue<DisplayOptions>("none");

  const handleFinish = () => {
    display.value = "none";

    if (onFinish) {
      onFinish();
    }
  };

  const Characters = useMemo(() => {
    const AnimatedName: React.JSX.Element[] = [];
    const name = String(children);
    display.value = "flex";

    for (let index = 0; index < name.length; index++) {
      AnimatedName.push(
        <Character
          {...props}
          start={start}
          onFinish={index === 0 ? handleFinish : undefined}
          index={index}
          key={`letter_${index}`}
          style={typeof style === "function" ? style(index) : style}
        >
          {name[index]}
        </Character>,
      );
    }

    return AnimatedName;
  }, [start, children, style]);

  const animatedContainer = useAnimatedStyle(() => ({
    display: display.value,
    flexDirection: "row",
  }));

  return (
    <Animated.View style={[animatedContainer, containerStyle]}>
      {Characters}
    </Animated.View>
  );
}
