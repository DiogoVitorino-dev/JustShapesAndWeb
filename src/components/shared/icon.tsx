import { Ionicons } from "@expo/vector-icons";
import React from "react";

type IconProps = React.ComponentProps<typeof Ionicons>;

export function Icon(icon?: IconProps) {
  return <Ionicons {...icon} />;
}
