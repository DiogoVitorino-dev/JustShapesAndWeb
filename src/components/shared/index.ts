export * from "./View";
export * from "./Text";
export * from "./icon";
export * from "./safeArea";
export * from "./headerBackButton";
export * from "./loading";
