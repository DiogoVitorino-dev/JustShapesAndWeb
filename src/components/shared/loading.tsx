import { LinearGradient } from "expo-linear-gradient";
import React, { useCallback, useEffect } from "react";
import { StyleSheet, useWindowDimensions } from "react-native";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withTiming,
} from "react-native-reanimated";

import { AnimatedText } from "./Text";

import Colors from "@/constants/Colors";
import { Position } from "@/constants/commonTypes";
import { MathUtils } from "@/utils/mathUtils";

interface ParticleProps extends Position {
  visibleDuration: number;
  size?: number;
}

const Particle = ({ x, y, size = 10, visibleDuration }: ParticleProps) => {
  const opacity = useSharedValue(0);

  useEffect(() => {
    opacity.value = withRepeat(
      withTiming(0.7, { duration: visibleDuration }),
      -1,
      true,
    );
  }, []);

  return (
    <Animated.View
      style={{
        opacity,
        width: size,
        height: size,
        top: y,
        left: x,
        position: "absolute",
        backgroundColor: Colors.loading.particle,
      }}
    />
  );
};

export function Loading() {
  const opacity = useSharedValue(1);
  const window = useWindowDimensions();

  useEffect(() => {
    opacity.value = withRepeat(withTiming(0.5, { duration: 500 }), -1, true);
  }, []);

  const createSquares = useCallback(() => {
    const { random } = MathUtils;
    const squares: React.JSX.Element[] = [];
    const quantity = window.width / 20;

    for (let index = 0; index < quantity; index++) {
      squares.push(
        <Particle
          size={random(3, 10)}
          visibleDuration={random(500, 2000)}
          x={random(0, window.width)}
          y={random(0, window.height)}
          key={`loadingSquare${index}`}
        />,
      );
    }

    return squares;
  }, [window]);

  const animatedStyle = useAnimatedStyle(() => ({ opacity: opacity.value }));

  return (
    <LinearGradient
      style={[styles.container]}
      colors={[Colors.loading.gradientStart, Colors.loading.gradientEnd]}
    >
      {createSquares()}
      <AnimatedText.Title style={[styles.text, animatedStyle]}>
        Loading
      </AnimatedText.Title>
    </LinearGradient>
  );
}

const styles = StyleSheet.create({
  container: {
    width: "100%",
    height: "100%",
    alignItems: "center",
    justifyContent: "center",
    position: "absolute",
  },

  text: {
    fontSize: 32,
  },
});
