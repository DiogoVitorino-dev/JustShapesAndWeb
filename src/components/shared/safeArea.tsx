import React from "react";
import {
  View as DefaultView,
  StyleProp,
  StyleSheet,
  ViewProps,
  ViewStyle,
} from "react-native";
import { useSafeAreaInsets } from "react-native-safe-area-context";

import Colors from "@/constants/Colors";

interface SafeAreaViewProps extends ViewProps {
  transparent?: boolean;
}

export const SafeArea = ({
  style,
  transparent,
  children,
  ...others
}: SafeAreaViewProps) => {
  const insets = useSafeAreaInsets();
  const styleArray: StyleProp<ViewStyle> = [styles.default];
  const contentStyleArray: StyleProp<ViewStyle> = [
    { ...insets },
    styles.default,
    style,
  ];

  if (transparent) {
    styleArray.push(styles.transparent);
    contentStyleArray.push(styles.transparent);
  }

  return (
    <DefaultView style={styleArray}>
      <DefaultView {...others} style={contentStyleArray}>
        {children}
      </DefaultView>
    </DefaultView>
  );
};

const styles = StyleSheet.create({
  default: {
    backgroundColor: Colors.UI.background,
    flex: 1,
  },

  transparent: { backgroundColor: "transparent" },
});
