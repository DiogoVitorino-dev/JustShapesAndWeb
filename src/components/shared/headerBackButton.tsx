import { NavigationProp } from "@react-navigation/native";
import React from "react";
import { StyleSheet, Pressable, View, Text } from "react-native";

import { Icon } from "./icon";

import Colors from "@/constants/Colors";
import { StackScreenOptions } from "@/constants/commonTypes";

type NativeHeaderBackButton = Exclude<
  Exclude<Exclude<StackScreenOptions, Function>, undefined>["headerLeft"],
  undefined
>;

type RouterParams = Parameters<Extract<StackScreenOptions, Function>>[0];
type HeaderBackButtonParams = Parameters<NativeHeaderBackButton>[0];

type Navigation = NavigationProp<ReactNavigation.RootParamList>;

export interface HeaderBackButtonProps
  extends HeaderBackButtonParams,
    RouterParams {
  onPress?: () => Awaited<Promise<void>>;
}

export function HeaderBackButton({
  route,
  onPress,
  canGoBack,
  label,
  tintColor,
  ...others
}: HeaderBackButtonProps) {
  const navigation: Navigation = others.navigation;

  const handlePress = async () => {
    if (canGoBack) {
      if (onPress) await onPress();

      navigation.goBack();
    }
  };

  return (
    <Pressable onPress={handlePress}>
      {({ pressed }) => (
        <View>
          <Icon
            style={[
              { transform: [{ scale: pressed ? 0.95 : 1 }] },
              styles.spacing,
            ]}
            name="arrow-back"
            size={26}
            color={tintColor}
          />
          {label ? (
            <Text style={[styles.label, styles.spacing]}>{label}</Text>
          ) : null}
        </View>
      )}
    </Pressable>
  );
}

const styles = StyleSheet.create({
  label: {
    fontFamily: "Manjari",
    fontSize: 16,
    color: Colors.UI.text,
  },

  spacing: {
    margin: 3,
    marginLeft: 12,
    padding: 3,
  },
});
