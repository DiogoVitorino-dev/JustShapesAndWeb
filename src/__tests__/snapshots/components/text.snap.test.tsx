import { render } from "@testing-library/react-native";

import { Text } from "@/components/shared";

describe("Text - Snapshot component test", () => {
  it("should render correctly", () => {
    const tree = render(<Text />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
