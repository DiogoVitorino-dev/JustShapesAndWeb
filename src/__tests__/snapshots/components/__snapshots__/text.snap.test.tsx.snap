// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text - Snapshot component test should render correctly 1`] = `
<Text
  style={
    [
      {
        "color": "rgb(245, 245, 245)",
        "fontSize": 22,
      },
      [
        undefined,
        {
          "fontFamily": "Manjari",
          "marginBottom": -3,
        },
      ],
    ]
  }
/>
`;
